extensions/sql-database-projects/src/models/project.ts

```diff
@@ -1106,6 +1113,47 @@ export class Project implements ISqlProject {
 
 		return firstPropertyElement.childNodes[0].data;
 	}
+
+	/**
+	 * Adds folder items for all intermediate folders in the path.
+	 * For example, when adding a file `a/b/c/d.sql`, we also need to add items for
+	 * the following folders:
+	 *   a/
+	 *   a/b/
+	 *   a/b/c/
+	 *
+	 * @param absolutePath Normalized absolute path to the file or folder.
+	 */
+	private async ensureFolderItemsForPath(absolutePath: string): Promise<void> {
+		const normalizedProjectFolderPath = path.normalize(this.projectFolderPath);
+
+		// Only add folders for files within the project folder. When adding files outside the project
+		// folder, they should be copied to the project root and there will be no additional folders to add.
+		if (!absolutePath.toUpperCase().startsWith(normalizedProjectFolderPath.toUpperCase())) {
+			return;
+		}
+
+		const relativePath = utils.convertSlashesForSqlProj(absolutePath.substring(normalizedProjectFolderPath.length));
+		const pathSegments = utils.trimChars(relativePath, ' \\').split('\\');
+		let intermediateFolderPath = '';
+
+		// Add folder items for all segments, except the last one, which will be explicitly
+		// added by the caller of this helper method
+		for (let segmentIndex = 0; segmentIndex < pathSegments.length - 1; ++segmentIndex) {
+			if (pathSegments[segmentIndex] !== '') {
+				intermediateFolderPath += pathSegments[segmentIndex] + '\\';
+				const existingEntry =
+					this.files.find(f => utils.ensureTrailingSlash(f.relativePath.toUpperCase()) === intermediateFolderPath.toUpperCase());
+
+				if (!existingEntry) {
+					// If there is no <Folder/> item for the intermediate folder - add it
+					const folderEntry = this.createFileProjectEntry(intermediateFolderPath, EntryType.Folder);
+					this.files.push(folderEntry);
```

Comment:
can consider moving this creating and adding the folder entry to the project file to a common function that can be called both here and in `addFolderItem()` so it's easier to update in the future, in case something changes

Reply:
So what I wanted to do initially (the reason why `await`s end up missing) is to call `addFolderItem` from `addScriptItem`, that way we can handle this "hierarchy" creation in the `addFolderItem`, but I decided against it because `addFolderItem` does bunch of other stuff that I didn't want to repeat (normalize the path yet again, recreate the folder, etc.), so I ended up adding this helper method, where I can assume that path will be normalized already and it doesn't have to happen again.

I will do what you suggested and extract these into a separate helper method, but maybe I should just go back to my original implementation and do this entire thing in the `addFolderItem` and just call it from the `addScriptItem`? What do you think?

Reply:
`addFolderItem` and `addScriptItem` were added by @Benjin in the early days of this extension. Any preference on what to do here?

Reply:
If we're going the route that every parent folder of each file in the project needs to be explicitly listed (rather than just displaying them in the virtual tree), then I think this responsibility makes the most sense in `addFolderItem()`.  `addScriptItem()` can then just start with `addFolderItem(path.dir(scriptPath))` and everything's clean.

Reply:
If you open the project with missing intermediate folders in VS with SSDT, then it will add them in there and save the sqlproj file without any prompts. In my mind that means we have already went the route where these folders need to be explicitly listed, no? What am I missing here in "route determination"?

So my original implementation had it all routed through `addFolderItem`, but I didn't like it because it was doing the path normalization again. Having this helper that was generating the intermediate folders by string concatenation, instead of using `path` methods and calling `convertSlashesForSqlProj` every time looked way more efficient. The code is also a little bit more complicated than just calling it with `path.dirname`, as it tends to return `.`, `/`, etc. depending on the platform and we will need to special case these.

Let me simplify the code by removing exceptions for duplicate items and I'll revisit the `addFolderItem` to see if i can optimize the code there a little bit.